### PR TITLE
relay: create channel bindings and relay data

### PR DESCRIPTION
**Pull request by thomaseizinger**
Here is a short demo:

[Relay](https://github.com/firezone/firezone/assets/5486389/c0199294-70ca-47b4-90ae-2c96428bdb56)

You can run this locally using the `./run_smoke_test.sh` shell-script. It is not reliable enough yet to be used in CI but I used one if its outputs to make a regression test.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
Need to fix the tests and add some automated ones for the channel bindings. Once that is done, I'd like to merge this because it is  a really big PR already :)

**Comment by thomaseizinger** [comment]
The examples work (sometimes)! :D

Next step is writing some automated tests and make the examples reliable so that I can use them in CI as tests.

**Review by jamilbk — COMMENTED** [review]
LGTM, but I'll defer to @conectado for final review

**Comment by jamilbk** [comment]
Btw, updated your base to latest `cloud`.

**Comment by jamilbk** [comment]
Modified steps I used to get this to run on aarch64 (e.g. Apple M1) under Docker:

1. Clone thomas's fork
1. Add `aarch64-unknown-linux-musl` target to `rust/rust-toolchain.toml`
1. `docker run --rm -it -v $(pwd)/rust:/root rust:alpine`
1. Then you can follow the steps inside that container:
```shell
apk add musl-dev
cargo build --target aarch64-unknown-linux-musl
RUST_LOG=relay=trace RELAY_LISTEN_IP4_ADDR=127.0.0.1 RELAY_PUBLIC_IP4_ADDR=127.0.0.1 cargo run --target aarch64-unknown-linux-musl --bin relay &
cargo run --target aarch64-unknown-linux-musl gateway &
cargo run --target aarch64-unknown-linux-musl client &
```

Still getting `Connection refused` though when running the gateway:

```
     Running `target/aarch64-unknown-linux-musl/debug/examples/gateway`
2023-05-20T19:26:32.347454Z TRACE relay: Received binding request from 127.0.0.1:53396
2023-05-20T19:26:32.347507Z TRACE relay: Sending binding success response to 127.0.0.1:53396
Our external address is 127.0.0.1:53396
Error: Connection refused (os error 111)
```

Similar behavior running natively on macOS. I'm sure getting this running in Docker on GitHub CI should iron things out. It would be nice to be able to test locally with Docker Desktop on macOS/Windows (aarch64/x86_64)

Nice work!

**Comment by thomaseizinger** [comment]
Yeah it doesn't work with docker yet. I wasted about 3 hours trying to set up a docker-compose script but for some reason, it is still way to hard to get proper caching to work for docker images.

Hence I gave up on that for now. We are going to deploy this on Linux only anyway right? So I think having that tested in CI is a good starting point.

Have you tried running it locally without docker? Don't see a reason why that shouldn't work on Mac.

**Comment by jamilbk** [comment]
> Have you tried running it locally without docker? Don't see a reason why that shouldn't work on Mac.

The relay started just fine, but the gateway and client both exit with Connection refused for some reason -- the same error I was getting in Docker (perhaps to no surprise). I tried with the macOS firewall both on and off with no change. In any case I agree with not wasting time getting it running on non-Linux platforms.

Once CI is working (and aarch64 target added) I'm sure it won't be a big hassle to test locally on Win/Mac/Linux via Docker Desktop.

File: rust/relay/src/main.rs
Comment (conectado) [inline rust/relay/src/main.rs:125]: But this also means we can't forward data nor handle timeouts as long as this socket is busy, right?

File: rust/relay/src/main.rs
Comment (conectado) [inline rust/relay/src/main.rs:81]: I'm worried that with this approach instead of using `tokio::select` one of the branches might hog this loop and we never make any progress with any of the other branches.

File: rust/relay/src/main.rs
Comment (conectado) [reply rust/relay/src/main.rs:125]: That also means that network buffers might be filled up while we wait for this socket to be ready since we are not reading incoming packets.

File: rust/relay/src/server.rs
Comment (conectado) [inline rust/relay/src/server.rs]: I think this could be a warn log.

File: rust/relay/src/udp_socket.rs
Comment (conectado) [inline rust/relay/src/udp_socket.rs:8]: I feel like it'd be less code for us to pass the buffer to the `EventLoop` rather than having this wrapper.

**Review by conectado — COMMENTED** [review]
This looks really good! I am glad it seems to be working(haven't tested it myself yet)

Left some comments to discuss but nothing major.

File: rust/relay/src/main.rs
Comment (thomaseizinger) [reply rust/relay/src/main.rs:81]: I am not sure if what I am doing is the right strategy here because we don't have backpressure in UDP anyway but the goal was to not have our internal buffers grow unbounded. We potentially have many UDP sockets (one per allocation) and one main socket. If I just take data from the allocation sockets whenever it is ready, then we may get unbounded buffer growth if our main socket cannot keep up with writing.

Without a concept of backpressure, we have to drop packets somewhere. In this implementation, that will happen likely in the network driver somewhere.

File: rust/relay/src/udp_socket.rs
Comment (thomaseizinger) [reply rust/relay/src/udp_socket.rs:8]: I went back and forth about 3 times but ended up settling on this because it is more convenient to use and I need it in two places (allocation and main socket). Feel free to experiment yourself!

**Comment by thomaseizinger** [comment]
> > Have you tried running it locally without docker? Don't see a reason why that shouldn't work on Mac.
> 
> The relay started just fine, but the gateway and client both exit with Connection refused for some reason -- the same error I was getting in Docker (perhaps to no surprise). I tried with the macOS firewall both on and off with no change. In any case I agree with not wasting time getting it running on non-Linux platforms.
> 
> Once CI is working (and aarch64 target added) I'm sure it won't be a big hassle to test locally on Win/Mac/Linux via Docker Desktop.

@jamilbk Did you start the redis container? That one is required as a signalling server (i.e. the job that the portal will fulfill).

File: rust/relay/src/server.rs
Comment (thomaseizinger) [reply rust/relay/src/server.rs]: Sure. The judgement I usually apply for `warn` is "Do I want to wake an ops person at 3am if this happens?".

This really only happens if someone is spamming our UDP socket with random data. Yeah, it might be a DoS but I think realistically, it is probably just random internet traffic probing ports.

File: rust/relay/src/main.rs
Comment (thomaseizinger) [reply rust/relay/src/main.rs:125]: Let's discuss this in the conversation below.

**Comment by thomaseizinger** [comment]
This is now passing tests. I also added a short shell-script that you can use to run it locally. Requires docker to be installed and running.

I am no longer very happy with the design of the regression tests. I'll likely adapt them to smaller unit tests that don't depend on wire input because that is too cumbersome and nobody can read that anyway unless you natively speak HEX-encoding :)

But that is for a future PR. I think this is a good baseline to merge.

File: rust/relay/src/server.rs
Comment (conectado) [reply rust/relay/src/server.rs]: Haha, gotcha, I was thinking of `warn` because it could be a meaningful log when debugging why a connection isn't working.

File: rust/relay/src/main.rs
Comment (conectado) [reply rust/relay/src/main.rs:81]: Hm I see what you mean. We can leave it like this and revisit this when we implement bandwidth, I still don't think we should let one of the branches "freeze" the event-loop.

Perhaps we can just use a ring buffer and drop received packets ourselves if a situation like that arises.

Comment (thomaseizinger) [reply rust/relay/src/main.rs:81]: > I still don't think we should let one of the branches "freeze" the event-loop.

The general pattern has been working really well for us in `rust-libp2p` but there we (aim to) have back-pressure between all the components. Maybe that is not a good idea in UDP-based systems?

> Perhaps we can just use a ring buffer and drop received packets ourselves if a situation like that arises.

Before we implement something like that, I would like to instrument the code with metrics and see how often that actually happens.

**Comment by jamilbk** [comment]
> > > Have you tried running it locally without docker? Don't see a reason why that shouldn't work on Mac.
> > 
> > 
> > The relay started just fine, but the gateway and client both exit with Connection refused for some reason -- the same error I was getting in Docker (perhaps to no surprise). I tried with the macOS firewall both on and off with no change. In any case I agree with not wasting time getting it running on non-Linux platforms.
> > Once CI is working (and aarch64 target added) I'm sure it won't be a big hassle to test locally on Win/Mac/Linux via Docker Desktop.
> 
> @jamilbk Did you start the redis container? That one is required as a signalling server (i.e. the job that the portal will fulfill).

I did, but perhaps I needed to forward ports to expose it to the gateway/client on localhost. I'll give it another test after this is merged.

**Review by conectado — APPROVED** [review]
Haven't run the smoke tests yet (I'm having problems with  docker + nix-darwin) but I think it's good to merge, can fix those later.

**Comment by jamilbk** [comment]
@thomaseizinger I think GH is complaining that some commits aren't signed. I'll override for now but you may want to double-check (unless the rebase somehow messed with it)

**Comment by jamilbk** [comment]
Aaayyyy getting promising output on macOS using `run_smoke_tests.sh`:

```
[  relay] 2023-05-23T00:32:05.590102Z  INFO relay: Listening for incoming traffic on UDP port 3478
[  relay] 2023-05-23T00:32:05.708230Z  INFO relay: Created new allocation for 127.0.0.1:54520 on 127.0.0.1:53432 and lifetime 600s
[ client] Allocated relay address: 127.0.0.1:53432
[gateway] Our external address is 127.0.0.1:53077
[ client] Retrieved gateway address: 127.0.0.1:53077
[ client] Created channel to gateway
[  relay] 2023-05-23T00:32:05.825148Z  INFO relay: Bound channel 16384 between 127.0.0.1:54520 and 127.0.0.1:53077 on allocation AID-1
[gateway] Client's relay address is 127.0.0.1:53077
[gateway] Sent ping to client: a77e490114e9813d0ea8a5ee59bdcb80416a9606281577430f414dd44f708a5c
[ client] Pushed relay address to gateway
[  relay] 2023-05-23T00:32:05.827380Z DEBUG relay: Relaying 32 bytes from 127.0.0.1:53077 to 127.0.0.1:54520 via channel 16384
[ client] Received buffer from 127.0.0.1:53077: a77e490114e9813d0ea8a5ee59bdcb80416a9606281577430f414dd44f708a5c
[  relay] 2023-05-23T00:32:05.827825Z DEBUG relay: Relaying 32 bytes from 127.0.0.1:54520 to 127.0.0.1:53077 via channel 16384
[gateway] Received pong from client: a77e490114e9813d0ea8a5ee59bdcb80416a9606281577430f414dd44f708a5c
[gateway] Sent ping to client: 76c5b3166c7b1bc060f4f0f3f5a87a658b8bb03638f8dcbd658d89981342de2b
[  relay] 2023-05-23T00:32:05.880402Z DEBUG relay: Relaying 32 bytes from 127.0.0.1:53077 to 127.0.0.1:54520 via channel 16384
[ client] Received buffer from 127.0.0.1:53077: 76c5b3166c7b1bc060f4f0f3f5a87a658b8bb03638f8dcbd658d89981342de2b
[  relay] 2023-05-23T00:32:05.880908Z DEBUG relay: Relaying 32 bytes from 127.0.0.1:54520 to 127.0.0.1:53077 via channel 16384
[gateway] Received pong from client: 76c5b3166c7b1bc060f4f0f3f5a87a658b8bb03638f8dcbd658d89981342de2b
[gateway] Sent ping to client: 838f0f48be69118f0a3237ecc2c748961ee24662ed353b7ddaea853dcc27b2a1
[  relay] 2023-05-23T00:32:05.933642Z DEBUG relay: Relaying 32 bytes from 127.0.0.1:53077 to 127.0.0.1:54520 via channel 16384
[ client] Received buffer from 127.0.0.1:53077: 838f0f48be69118f0a3237ecc2c748961ee24662ed353b7ddaea853dcc27b2a1
[  relay] 2023-05-23T00:32:05.934124Z DEBUG relay: Relaying 32 bytes from 127.0.0.1:54520 to 127.0.0.1:53077 via channel 16384
[gateway] Received pong from client: 838f0f48be69118f0a3237ecc2c748961ee24662ed353b7ddaea853dcc27b2a1
[gateway] Sent ping to client: 2f7aa0582046c0128d33ce7847e99aced99a45b4d2d9dafeb8be80498a5e96c5
[  relay] 2023-05-23T00:32:05.986780Z DEBUG relay: Relaying 32 bytes from 127.0.0.1:53077 to 127.0.0.1:54520 via channel 16384
[ client] Received buffer from 127.0.0.1:53077: 2f7aa0582046c0128d33ce7847e99aced99a45b4d2d9dafeb8be80498a5e96c5
[  relay] 2023-05-23T00:32:05.987231Z DEBUG relay: Relaying 32 bytes from 127.0.0.1:54520 to 127.0.0.1:53077 via channel 16384
[gateway] Error: deadline has elapsed
./relay/run_smoke_test.sh: line 6: 69995 Terminated: 15          $client 2>&1
     69996                       | sed "s/^/${RED}[ client]${NC} /"
./relay/run_smoke_test.sh: line 6: 69997 Terminated: 15          $relay 2>&1
     69998                       | sed "s/^/${GREEN}[  relay]${NC} /"
```

**Comment by thomaseizinger** [comment]
> [gateway] Error: deadline has elapsed

I doesn't work every time unfortunately but you can try run it multiple times :)

**Comment by thomaseizinger** [comment]
> @thomaseizinger I think GH is complaining that some commits aren't signed. I'll override for now but you may want to double-check (unless the rebase somehow messed with it)

Yep, that is due to rebasing. Rebasing changes the parent hash so the commit hash changes and thus the signature is void.

**Pull request merged**
